src/common/walFilter/walFilter.c

```diff
@@ -0,0 +1,763 @@
+    MEM_CONTEXT_TEMP_BEGIN();
+
+    const String *walSegment = NULL;
+    const TimeLineID timeLine = main_state->current_header->xlp_tli;
+    uint64_t segno = main_state->current_header->xlp_pageaddr / main_state->segSize;
+    const String *const path = strNewFmt("%08X%08X", timeLine, (uint32) (segno / XLogSegmentsPerXLogId(main_state->segSize)));
+
+    const StringList *const segmentList = storageListP(
+        storageRepoIdx(main_state->archiveInfo->repoIdx),
+        strNewFmt(STORAGE_REPO_ARCHIVE "/%s/%s", strZ(main_state->archiveInfo->archiveId), strZ(path)),
+        .expression = strNewFmt("^[0-f]{24}-[0-f]{40}" COMPRESS_TYPE_REGEXP "{0,1}$"));
+
+    if (strLstEmpty(segmentList))
+    {
+        // an exotic case where we couldn't even find the current file.
+        THROW(FormatError, "no WAL files were found in the repository");
+    }
+
+    uint64_t segno_diff = UINT64_MAX;
+    for (uint32_t i = 0; i < strLstSize(segmentList); i++)
+    {
+        const String *const file = strSubN(strLstGet(segmentList, i), 0, 24);
+        TimeLineID tli;
+        uint64_t fileSegNo = 0;
+        XLogFromFileName(strZ(file), &tli, &fileSegNo, main_state->segSize);
+
+        if (segno - fileSegNo < segno_diff && fileSegNo < segno)
+        {
+            segno_diff = segno - fileSegNo;
+            walSegment = strLstGet(segmentList, i);
+        }
+    }
+
+    // current file is oldest
+    if (segno_diff == UINT64_MAX)
+    {
+        memContextSwitchBack();
+        memContextDiscard();
+        return false;
+    }
+
+    const StorageRead *const storageRead = storageNewReadP(
+        storageRepoIdx(0),
+        strNewFmt(STORAGE_REPO_ARCHIVE "/%s/%s", strZ(main_state->archiveInfo->archiveId), strZ(walSegment)));
+
+    buildArchiveGetPipeLine(ioReadFilterGroup(storageReadIo(storageRead)), main_state->archiveInfo);
+
+    OBJ_NEW_BEGIN(WalFilterState, .childQty = MEM_CONTEXT_QTY_MAX, .allocQty = MEM_CONTEXT_QTY_MAX)
+    {
+        memset(this, 0, sizeof(WalFilterState));
+        this->is_begin = true;
+        this->record = memNew(this->heapPageSize);
+        this->rec_buf_size = this->heapPageSize;
+        this->headers = lstNewP(SizeOfXLogLongPHD);
+        this->walPageSize = main_state->walPageSize;
+        this->heapPageSize = main_state->heapPageSize;
+        this->segSize = main_state->segSize;
+
+        this->walInterface = &interfaces[0];
+    }
+    OBJ_NEW_END();
+
+    ioReadOpen(storageReadIo(storageRead));
+
+    Buffer *const buffer = bufNew(this->walPageSize);
+    size_t size = ioRead(storageReadIo(storageRead), buffer);
+    bufUsedSet(buffer, size);
+
+    while (true)
+    {
+        if (!readRecord(this, buffer))
+        {
+            if (ioReadEof(storageReadIo(storageRead)))
+            {
+                if (this->got_len >= offsetof(XLogRecord, xl_rmid) + SIZE_OF_STRUCT_MEMBER(XLogRecord, xl_rmid))
+                {
+                    // xl_info and xl_rmid is in prev file. Nothing to do
```

Comment:
maybe in next file?

Reply:
No, that's right. This function reads the beginning of the recording from the previous file. And if the fields we need are in the previous file, then we have nothing to do with this record at the moment.